### ISSUE TITLE
Realt-time collaboration

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-toolpad#418]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

Make it possible to collaborate on the same page.

### Examples 🌈

- https://wishlist.webflow.com/ideas/WEBFLOW-I-39
- https://liveblocks.io/
- https://github.com/ToolJet/ToolJet/pull/2517
- https://vercel.com/live

### Motivation 🔦

This is similar to #348 but covers multiple editors. To be honest, seeing Webflow, Retool not having this feature, and how developers work in their own PRs, I doubt this is a feature developer would really want. But who knows, upvotes will tell us.

**Comment by Janpot** [comment]
I strongly believe most people look the wrong way at what's the benefit of real time collaboration. The big advantage really isn't about working in a document simultaneously. It's about having a single source of truth. About knowing that whenever you open the document, you're looking at the latest version, and having the confidence that when you're editing a document, your changes will be in there for the next person. It's about avoiding branching off and having to maintain multiple versions of the same document.
The fact that I see someone else's cursor move and realtime type is a nice gimmick but most of the time it's not practical working on a document simultaneously.

**Comment by oliviertassinari** [comment]
> About knowing that whenever you open the document, you're looking at the latest version

Strong agree with this. Same in the GitHub PRs view that has a websocket to let you know that there is a new commit.

> Tooljet, dynaboard and appsmith have it.

I would be careful with this. Maybe they got trapped into the same thinking as of https://medium.com/@christian.dobbert/the-missing-bullet-holes-and-abraham-wald-25e68d7a870f at 💯. "Retool doesn't have X, let's add X". I think that it should be more: "What is a core dimension of the problem that we can make x10 better?"

**Comment by Janpot** [comment]
> Strong agree with this. Same in the GitHub PRs view that has a websocket to let you know that there is a new commit.

Yes, but merging conflicts manually is still a difficult process. the 10x better approach might as well be, "eliminate the possibility for merge conflicts altogether".

The implementations in appsmith and tooljet feel like they're an afterthought. They are indeed bad examples. Maybe we should rather look at google docs vs. word. Or figma vs. sketch? I may be wrong but I have a sense that their realtime nature is a great contributor to their success. Maybe the paradigm is only useful for editing documents?

**Comment by newguy-123** [comment]
> Maybe the paradigm is only useful for editing documents?

I guess so. A design/document has a lower threshold for anyone to contribute, hence collaboration is needed. But the same doesn't apply to building a production app. 
In the broader view of allowing collaboration. I see 3 options:

1. Real-time commenting: A user will be able to see others commenting and their pointer moving. Retool doesn't have this. Though it is possible through Cord: https://retool.com/blog/build-together-in-retool-with-cord-comments-and-annotations/
2. Multi-user editing: Like the present Toolpad, Retool experience, and it is not real-time. https://community.retool.com/t/multiples-users-collaborating-on-the-same-app/3852
3. Real-time multi-user editing: As discussed above like Figma/Google Docs 

I think before we consider 3. we should do 1.
